design/ceph/cephx-keyring-rotation.md

```diff
@@ -0,0 +1,529 @@
+keyring. It will be created to rotate the primary admin keyring and removed after rotation.
+
+While `client.admin-rotator` exists, it must be stored in the secret also. The secret's new data
+format will look as follows:
+
+- `data`
+    - `keyring`: contains the primary `client.admin` keyring
+    - `rotatorKeyring`: if present, contains the temporary `client.admin-rotator` keyring
+
+### Rotation procedure
+
+First, we establish a rotation procedure. Because this procedure is risky, the process takes extra
+caution to verify that updates to the secret are properly stored. There is no room for error.
+
+1.  as `client.admin`: run `ceph auth get-or-create client.admin-rotator` w/ admin permissions
+2.  store/update the `client.admin-rotator` key into the secret `data.rotatorKeyring`
+3.  read back `client.admin-rotator` keyring from the secret, and verify it
+4.  **ROTATE**: as `client.admin-rotator`: run `ceph auth ls` to ensure it has permissions
+5.  as `client.admin-rotator`: run `ceph auth rotate client.admin`
+6.  as `client.admin` run `ceph auth ls` to ensure it has permissions
+7.  update on-disk ceph config file with new `client.admin` keyring
+8.  update the secret `data.keyring` with new `client.admin` keyring
+9.  read back `client.admin` keyring from the secret, and verify it
+10. as `client.admin`: run `ceph auth rm client.admin-rotator`
+11. remove `data.rotatorKeyring` from the secret
+12. update `CephCluster.status.cephx.admin` with updated CephX status
+
+### In post-mon-startup actions
+
+Rotation of the admin key should happen after mons are updated. This is important if Ceph is being
+upgraded and the 'current' Ceph version doesn't support the `auth rotate` command, but the 'new'
+Ceph version does support it. (This consideration also exists for rotating the `mon.` key)
+
+1. use `CephCluster.spec.security.cephx.daemon` to determine if rotation is indicated
+2. if indicated, perform the rotation procedure from start to finish
+
+### In CephCluster startup
+
+If rotation fails or the operator restarts after the **ROTATE** step, the CephCluster reconcile will
+begin again. In the new reconcile, the secret may have the wrong info in `data.keyring`. In that
+case, the CephCluster reconcile would be unable to take any admin actions, effectively bricking the
+reconcile. Therefore, the CephCluster reconciler must be able to recover from any interrupted admin
+key rotation and must do so before any other Ceph admin actions.
+
+1. read the secret (same as today)
+    1. if secret `data.rotatorKeyring` is not present, no recovery needed
+        1. load `client.admin` from the secret `data.keyring` and continue reconcile (same as today)
+    2. if secret `data.rotatorKeyring` is present, prior rotation failed somewhere -- now recover
+        1. load `client.admin-rotator` from the secret `data.rotatorKeyring`
+        2. GOTO: **ROTATE**
```

Comment:
I need to update the design to handle a corner case here: if `data.rotatorKeyring` is present **but** step 10 (run `ceph auth rm client.admin-rotator`) already ran, `client.admin-rotator` won't exist in mon auth. In this case, we should shortcut directly to step 11.